Development/nmos/control_protocol_methods.cpp

```diff
@@ -0,0 +1,550 @@
+                {
+                    return make_control_protocol_response(handle, { nc_method_status::read_only });
+                }
+
+                if ((val.is_null() && !nmos::fields::nc::is_nullable(property))
+                    || (val.is_array() && !nmos::fields::nc::is_sequence(property)))
+                {
+                    return make_control_protocol_response(handle, { nc_method_status::parameter_error });
+                }
+
+                resources.modify(resource, [&](nmos::resource& resource)
+                {
+                    resource.data[nmos::fields::nc::name(property)] = val;
+
+                    resource.updated = strictly_increasing_update(resources);
+                });
+                return make_control_protocol_response(handle, { nc_method_status::ok });
+            }
+
+            // unknown property
+            utility::stringstream_t ss;
+            ss << U("unknown property: ") << property_id.serialize() << " to do Set";
+            return make_control_protocol_error_response(handle, { nc_method_status::property_not_implemented }, ss.str());
+        }
+
+        // Get sequence item
+        web::json::value get_sequence_item(nmos::resources& resources, nmos::resources::iterator resource, int32_t handle, const web::json::value& arguments, get_control_protocol_class_handler get_control_protocol_class, get_control_protocol_datatype_handler, slog::base_gate& gate)
+        {
+            // note, model mutex is already locked by the outter function, so access to control_protocol_resources is OK...
+
+            const auto& property_id = nmos::fields::nc::id(arguments);
+            const auto& index = nmos::fields::nc::index(arguments);
+
+            slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Get sequence item: " << property_id.serialize() << " index: " << index;
+
+            // find the relevant nc_property_descriptor
+            const auto& property = find_property(parse_nc_property_id(property_id), parse_nc_class_id(nmos::fields::nc::class_id(resource->data)), get_control_protocol_class);
+            if (!property.is_null())
+            {
+                if (!nmos::fields::nc::is_sequence(property))
+                {
+                    // property is not a sequence
+                    utility::stringstream_t ss;
+                    ss << U("property: ") << property_id.serialize() << U(" is not a sequence to do GetSequenceItem");
+                    return make_control_protocol_error_response(handle, { nc_method_status::invalid_request }, ss.str());
+                }
+
+                auto& data = resource->data.at(nmos::fields::nc::name(property));
+
+                if (!data.is_null() && data.as_array().size() > (size_t)index)
+                {
+                    return make_control_protocol_response(handle, { nc_method_status::ok }, data.at(index));
+                }
+
+                // out of bound
```

Comment:
Maybe "out of bounds" can be misleading: if `data.is_null()` returns true this also an error but not out of bounds?
See also lines 152 and 245.

Reply:
Thanks, well spotted.